### ISSUE TITLE
Attempting to validate nodeKind sh:IRI is useless  

### ISSUE THREAD

**Issue by umbreak** [INCF/neuroshapes#161]
Validations on `property` with `"nodeKind": "sh:IRI"` with `"minCount": 0` don't work.

For example, [this test](https://github.com/INCF/neuroshapes/blob/master/modules/nexus-schemas/src/test/resources/invalid/nexus/core/mediatype/v0.1.0/non-IRI-isDefinedBy.json#L6) against [this schema](https://github.com/INCF/neuroshapes/blob/master/modules/nexus-schemas/src/main/resources/schemas/nexus/core/mediatype/v0.1.0.json#L43) passes because the workbench signals the wrong message.

The error happens before the validator kicks in, when trying to convert the triples into a Jena Model: `Exception: Illegal character in path at index 5: wrong IRI`. This is Just Jena saying that a field expected to have a @type of @id cannot have this character (space).

If you had tried to put `"isDefinedBy":"wrong"` on the data to test, the test would pass (even though we don't want it to pass, since `wrong` is not an IRI. This is due to the conversion from JSON-LD to Jena Model.

If we have something like this:

```json
{
  "@context": {
    "dcterms": "http://example.com/",
    "isDefinedBy": {
      "@id": "dcterms:isDefinedBy",
      "@type": "@id"
    }
  },
  "@type":"dcterms:FileFormat",
  "isDefinedBy":"wrong"
}
```
when converting it to Jena (without having any `base`), the triple `isDefinedBy` gets removed and it will never be taken into account for the validation step. This is because `isDefinedBy` does not get expanded to `dcterms:isDefinedBy` because it does not match the rule `"@type": "@id"`.

**Comment by MFSY** [comment]
Hi,
Thank you for looking at this.
I would not say that **"Attempting to validate nodeKind sh:IRI is useless"**. I think it is a too strong statement here.  

checking that a node is an IRI through "nodeKind":"sh:IRI" is definitely usefull and I believe that's why it was added to the SHACL core spec.

But I think you want to point that depending on how the user write his/her json-ld document, some triples may reach the validator or not being ignore when deserialized.

I have few comments on that:

1. For the user, there is only one validator: the user does not know if the shacl validator is based on Jena or not. So in that case, the process of loading the json-ld payload into a Jena model prior to send it to a SHACL validator is completely transparent and corresponds to an atomic operation (I want to validate my json-ld payload).

2. Let take the following property shape (which corresponds to the one you mentioned above):

```json
{
          "path": "rdfs:isDefinedBy",
          "name": "Defined by",
          "nodeKind": "sh:IRI",
          "maxCount": 1
        }
```
The following data should fail because "mybase/wrong" is not a correct IRI:

```json
{
  "@context": {
    "dcterms": "http://example.com/",
    "isDefinedBy": {
      "@id": "rdfs:isDefinedBy",
      "@type": "@id"
    },
"@base":"mybase/"
  },
  "isDefinedBy":"wrong"
}
```

The following data should pass:

```json
{
  "@context": {
    "dcterms": "http://example.com/",
    "isDefinedBy": {
      "@id": "rdfs:isDefinedBy",
      "@type": "@id"
    },
"@base":"http://goodbase"
  },
  "isDefinedBy":"wrong"
}
```

The following data should fail because rdfs:isDefinedBy property is present with a value which is not an IRI (again this is a user perspective). The fact that "isDefinedBy" property is ignored or not when a @base is present or not is an implementation detail. If you look at the following data in the [json-ld playground](https://json-ld.org/playground/#startTab=tab-expanded&json-ld=%7B%22%40context%22%3A%7B%22dcterms%22%3A%22http%3A%2F%2Fexample.com%2F%22%2C%22isDefinedBy%22%3A%7B%22%40id%22%3A%22dcterms%3AisDefinedBy%22%2C%22%40type%22%3A%22%40id%22%7D%7D%2C%22%40type%22%3A%22dcterms%3AFileFormat%22%2C%22isDefinedBy%22%3A%22wrong%22%7D) then you'll see that a default @base (https://json-ld.org/playground/) is used whenever one is not asserted in the user provided payload. I would imagine [Nexus](https://github.com/BlueBrain/nexus-kg) implements such default base feature.

```json
{
  "@context": {
    "dcterms": "http://example.com/",
    "isDefinedBy": {
      "@id": "rdfs:isDefinedBy",
      "@type": "@id"
    }
  },
  "isDefinedBy":"wrong"
}
```

**Comment by umbreak** [comment]
Yes I fully agree. That's why I was raising awareness that right now, the last example you posted should failed **but it doesn't**. The current test, which has `"isDefinedBy":"wrong IRI"` fails, as expected but just because there is a space and Jena is able to catch that error.

I just wanted to clarify that the test is misleading because it seems to cover the case when isDefinedBy is not a URI, but it doesn't cover it.

And the way to fix this is by having some default @base , as you mentioned
